### ISSUE TITLE
Port to WebExtension

### ISSUE THREAD

**Issue by SammysHP** [vimperator/vimperator-labs#705]
Sören Hentzschel has published a roadmap regarding multiprocess architecture and addon compatibility: https://www.soeren-hentzschel.at/firefox/so-geht-es-weiter-mit-dem-multiprozess-firefox-2/

(See also https://blog.mozilla.org/addons/2016/11/23/add-ons-in-2017/)

Especially two releases are important for Vimperator:

**Firefox 53 (2017-04-18):** Only WebExtensions allowed on AMO. This means that new developers cannot create new unlisted addons for signing.

**Firefox 57 (2017-11-28):** Only WebExtensions allowed in Firefox and AMO. Vimperator will not work with this release anymore.

Please note that the developer version of Firefox might drop support for XUL- and SDK-addons already some months earlier, maybe in June!

I have a really bad feeling about this. Considering the development activity during the last months I don't think that we will have ported Vimperator to a WebExtension in time. This process requires almost a complete rewrite of Vimperator and a lot of features aren't possible with WebExtensions at all. Don't get me wrong, but I fear that this is the death of Vimperator. :(

See also #211 for some more details, which focusses on e10s compatibility in general.

## Chatroom

**I created a chatroom on Gitter where we can organize and talk about the next steps:**
**https://gitter.im/vim-webextension/Talk**

**Comment by hejops** [comment]
If all else fails, could focus perhaps be shifted to compatibility with Pale Moon?

**Comment by timss** [comment]
>I have a really bad feeling about this. Considering the development activity during the last months I don't think that we will have ported Vimperator to a WebExtension in time. This process requires almost a complete rewrite of Vimperator and a lot of features aren't possible with WebExtensions at all. Don't get me wrong, but I fear that this is the death of Vimperator. :(

It would indeed seem that the best possible and realistic outcome now would be to lag behind on Firefox releases for some time until a rewrite is completed, or even started. 

At some point it stops making sense to even maintain the current code base. I would hope that we could get a working release (especially on AMO) for the next ESR (FF52), which in my opinion would be a good option to stay on for some time as releases after 52 will be increasingly hard to maintain.

---

Regarding the option of Pale Moon, there's a pretty interesting discussion on that in this Pentadactyl issue. I don't personally think it's a good idea.

https://github.com/5digits/dactyl/issues/99

**Comment by svarogg** [comment]
I would really love to help with the effort. However, a few pointers on where can one start would be very helpful.

I've done the reading, and as I understand we are facing two major roadblocks:
1. New APIs means a complete or at least a very extensive rewrite.
2. Some APIs are not available, which makes some features hard or impossible to implement.

A couple of thoughts:
I think point No 1. warrants a new repository or at least a clean branch, especially if we are talking about a complete rewrite.

To address point No. 2 it would be nice to have a list of missing APIs, and might start thinking about workarounds.

I have much development experience, but none in browser extensions and haven't touched JS for a couple of years. So I don't think I am fit to lead this effort, but I would be more than happy to follow if pointed at the right direction.

**Comment by marza91** [comment]
Hey, I just started using Firefox and vimperator after being a long-time Chromium/Vimium-user (Google Amp on my phone slowly made me turn away from google). I love the experience of using gui=none so far and would love to contribute some with keeping this project alive!

That being said, I don't have any experience writing Firefox plugins besides building the latest xpi a few days ago, though I'm a pretty decent developer working mostly in web. I'm guessing I should start by reading a bit about writing plugins in general, would reading up on WebExtensions be the right place to start right now? And/or should I take a look at some open issues already around?

**Comment by Nefari0uss** [comment]
@marza91 If I had to guess, I'd say Web Extensions might be worth looking into. Someone is gonna have to start a port and the earlier the better is my likely guess.

**Comment by marza91** [comment]
This is probably possible to find out somewhere, but is Vimperator a "XUL/XPCOM" or "Add-ons SDK" extension?

**Comment by apheleia** [comment]
Count me in when moving to WebExtensions 👍

**Comment by SammysHP** [comment]
@svarogg
A new repository is a good idea. Yes, it would be a complete rewrite. With a new repository it would be easier to manage issues etc.

For the start we need some experienced software engineers to create an initial framework. Later individual developers can contribute additional features individually. There will be a long design phase where we collect requirements, develop a solid foundation for the addon and where we can also analyze which features are missing in the webextension standard.

@marza91
Vimperator is a XUL addon.

**Comment by SammysHP** [comment]
Oh, I forgot something: The new version would be something completely new. There will be many things that will be different from what we are used to. It would be fantastic to forget the reasons that lead to the fork and work together with the people from pentadactyl. They have the same problem as we have. And most developers joined after the fork. It makes no sense to create two new addon with the same functionality.

**Comment by cornop** [comment]
Personally, I will be switching to [VimFx](https://github.com/akhodakivskiy/VimFx). It doesn't change the UI, which is one thing I liked about Vimperator, but that won't be possible in the future anyway. It seems to be ahead of Vimperator in terms of Firefox's changes (it already supports multiprocess), is actively developed and [has a roadmap for WebExtension support](https://github.com/akhodakivskiy/VimFx/issues/860).

I'll just throw this in the room: Maybe the remaining developer power should be moved to this project.

**Comment by Dioxylin** [comment]
Regarding Pale Moon, their roadmap, at https://www.palemoon.org/roadmap.shtml, explicitly states that they are keeping XPCOM and XUL extensions support, unlike Firefox.

**Comment by marza91** [comment]
The ability to change the GUI and the way it completely changes how you use the browser is what I like most about Vimperator as well, so VimFx seems like a downgrade to me. Not to talk down about the Add-on itself, I bet it works great, it's just not exactly what I want (I tried it briefly before changing to Vimperator).

Would it be an idea to start by assembling a feature-list and find out exactly what can and can't be ported? I'm also lending a bit towards switching to Pale Moon if there's interest in keeping the XUL version going and most features can't be ported.

**Comment by gkatsev** [comment]
Collecting a feature list and categorizing it into "supported by webext" sounds like an awesome idea.

**Comment by xdml** [comment]
I follow your discussion about moving forward to WebExtensions age. I do not know exactly what is possible and what not, but from the discussion I got feeling, that interface rewrite by XUL will not be possible.

My question is, can be some of these issues solved by theme? So basically some XUL features will be done by theme?

**Comment by tfrangio** [comment]
I would like to throw my hat in the ring as a developer that can help with
the port. Do not have time to lead the project but as soon as we find a
leader(s) I can take on tasks.

**Comment by timss** [comment]
New repo would be practical due to the high number of unresolved issues in this repo, but it would also mean that one can't as easily refer to still active issues using #705. There's also lots of commit statistics, links to the repo on the Internet etc. As a minimum, the main repo would have to include information and links to the other one. 

That being said, I don't plan on manage the move to a new repo and/or branch, so it'll be up to whoever makes the call. Just thought I'd point out some possible pros/cons.

**Comment by lilydjwg** [comment]
You can easily refer to issues in other repos by pasting the full links, or use `user/repo#xxx` syntax.

**Comment by SammysHP** [comment]
Yes, cross-linking issues is easy. The new code would not have much in common with the old Vimperator code, so I don't see a problem. As we will start from scratch, most issues would have no relevance anymore.

**Comment by timss** [comment]
I'm well aware, @lilydjwg. My point is that there's some history that comes with all the issue numbers. In the event a new repo is created, an issue discussed with number e.g. 77 could be different, even though an issue 77 in the old repo could have historical value. Not a big problem, but it's not just about not having to write `user/repo#N` vs `#N`.

**Comment by fakedrake** [comment]
Because there was some talk about palemoon and other browsers. I switched temporarily to IceCat because vimperator still works properly there.

**Comment by SammysHP** [comment]
@timss I don't understand what you mean. The new WebExtension version of Vimperator would be a completely new program. There are no bugs, only feature requests (many more than there are now because it won't have any features at the beginning). So in the rare situation that it is necessary to reference an old issue it would be no problem to use the `repo#N` syntax.

**Comment by fakedrake** [comment]
It wasn't clear from my comment but I would agree with@SammysHP as the current version of vimperator should still work with current/legacy browsers like icecat and palemoon.

**Comment by timss** [comment]
Just thought I'd give my two cents, issue linking wasn't in itself a huge motivation. Please, don't let my remark become too much of a focus, I don't feel too strongly about it :)

**Comment by jrmiller82** [comment]
So.... firefox is doing something internally that's going to break vimperator? (Or already broke it? I'm on FF 51 and lots of things aren't working correctly anymore).

**Comment by Nefari0uss** [comment]
@jrmiller82 Mozilla is making the decision to move away from XUL for their extension framework in favour of web extensions. This move will ultimately make it more difficult to customize nearly as much of the browser and will require lots of extensions to be rewritten. Their reasoning is that the current extension system makes it difficult to update the browser with new features as maintaining compatibility becomes increasingly more difficult. I'm not going to discuss the politics or the pros and cons of this move (that's already been done to death) but the fact of the matter is that it's being done. This means that ultimately Vimperator is going to have to be rewritten essentially from scratch to work with web extensions. This means for a while it's going to be rather painful as either we won't be able to use the extension or that whatever version works with the new ecosystem will be very lite in terms of feature set. Ideally the growing pains of WebExtensions gets re-mediated over a period of time and we have a new ecosystem, a new code base to work with (with whatever improvements starting over can provide), and hopefully (near) feature parity in due time.

**Comment by fakedrake** [comment]
So maybe someone who knows the ins and outs of vimperator and has some idea of what are the priority features should break down what rewriting vimperator would entail so we can get going. I was looking at the [XUL/XPCOM vs WebExtensions article](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Comparison_with_XUL_XPCOM_extensions) and it looks like very few of the useful features of vimperator are translatable to WebExtensions.

**Comment by jrmiller82** [comment]
So in the meantime, is there a vim-like plugin that does work with the new WebExtensions? Or *gasp*, will I have to use Chrome and vimium or something awful like that? :-1:

**Comment by fakedrake** [comment]
I am currently using conkeror and IceCat+vimperator trying to decide
which one is a better short term solution. Conkeror is barely winning
for now...

**Comment by jrmiller82** [comment]
@Nefari0uss Forgot to say thanks for your detailed response. I wasn't trying to spark a political debate. I realize that change happens in the upstream always. :)

**Comment by Nefari0uss** [comment]
@jrmiller82 No problem! As of now I have VimFX on Firefox as a backup since Vimperator doesn't work with Nightly.

**Comment by jrmiller82** [comment]
@Nefari0uss I'll have to check that out.

**Comment by dschrempf** [comment]
Thanks for the reference to the discussion of the folks using pentadactyl.  I would like to 
add a reference to a related VimFX issue akhodakivskiy/VimFx#860.  I like the argument of the VimFX author.

I prefer Vimperator, because I cannot work with a command line that takes nearly half a second to show up but I believe it makes a lot of sense to join forces.

**Comment by SammysHP** [comment]
Firefox 57 will be released two weeks earlier: https://wiki.mozilla.org/index.php?title=RapidRelease%2FCalendar&diff=1162727&oldid=1160751

**Comment by pauldub** [comment]
Vimperator is the only reason I use Firefox, if a there is a call for help I would be happy to contribute. I have started thinking about this for some time and I do not know what I would do if vimperator stopped working. My requirements for a web browser are quite narrow and I even started writing a GUI using CEF with the goal of implementing only the minimum functionality I need from a browser. I'd rather contribute to porting Vimperator ;)

**Comment by maxauthority** [comment]
I fully agree with the others, that starting a Vimperator web-extension should happen in a new repository, and once we have some people being commited in doing that, we shall create a vimperator-web-extension repository in https://github.com/vimperator/ just like we did for the (potentially) new home page and syntax highligthing.

And while on the one hand it'll be sad to see the full XUL vimperator gone, on the other hand it'll free us from the Firefox backend, and people can choose freely between Firefox, Chrome, Edge or Opera, as they all support WebExtensions.

That said, it still requires *somebody* to do the first step and start a basic framework/skeleton, and be it just with very simple commands like h/j/k/l, and then adding features back in one by one. Probably it'll be easier to just use Vimium (for Chrome) as a base and make it work on Firefox.

**Comment by mkoskar** [comment]
@dschrempf 

> I prefer Vimperator, because I cannot work with a command line that takes nearly half a second to show up ...

I believe what you are talking about is built-in "Developer Toolbar" (Shift-F2). VimFx doesn't have its own command line but just focuses browser's urlbar.

**Comment by dschrempf** [comment]
@mkoskar 
Yes, this may be what I mean.  The bar that I am talking about shows up when I press `:`.

**Comment by SammysHP** [comment]
I can summarize what is most important for me:

- Commandline/statusbar
- <kbd>o</kbd>, <kbd>O</kbd>, <kbd>t</kbd>, <kbd>T</kbd>, <kbd>w</kbd>, <kbd>W</kbd> – they must open something like the completion in Vimperator.
- <kbd>f</kbd>, <kbd>F</kbd> and <kbd>;y</kbd> to follow and copy links
- <kbd>/</kbd> with the current behavior, the normal findbar doesn't work for me
- <kbd>y</kbd> and <kbd>p</kbd> to copy and paste URL
- Mappings that can override Firefox keybindings like <kbd>Ctrl</kbd>+<kbd>Q</kbd>
- <kbd>b</kbd> to show and switch buffers
- `:set gui=none` – just me, the content and a statusbar
- Some kind of NoScript integration. Showing the popup would be enough.
- Keyboard navigation of course (<kbd>h</kbd>, <kbd>j</kbd>, <kbd>k</kbd>, <kbd>l</kbd>, <kbd>gg</kbd>, <kbd>G</kbd>)
- Open an external editor (<kbd>Ctrl</kbd>+<kbd>I</kbd>) would be nice
- Commands to change preferences (`:set! …`)
- Tab commands (<kbd>d</kbd>, <kbd>u</kbd>, `:tabdetach`, `:tabmove`)

**Comment by xdml** [comment]
Has anyone checked https://github.com/shinglyu/QuantumVim?

Looks like @shinglyu already started similar project.

**Comment by shinglyu** [comment]
@xdml Thanks for mentioning. I've implemented a small set of features that is critical to my daily use, but I haven't put much effort into making it a consumer grade product. I'd appreciate if anyone can help by submitting PRs, bugs or feature requests.

**Comment by shinglyu** [comment]
FWIW, I found fa few issue during my implementation of [QuantumVim](https://github.com/shinglyu/QuantumVim):

* WebExt lacks a way to add global hotkey, you must add "Ctrl+" or "Shift+", and they sometimes conflict with the built-in hotkey. I have to use a page script to capture user hotkeys, but that's not very reliable.
* For features like `:open`, I just want to focus on the URL bar and select all. I already filed a [bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1295400) but for now I have to implement that in the Add-on SDK. So people have to install two add-ons, one in WebExt and one in Add-on SDK.

**Comment by xdml** [comment]
@shinglyu Do you think it will be possible to implement command line instead focusing address bar or use awesome bar as command line? My preference is to use `gui=none` in Vimperator, so willing to have command line in future as well. And that's why I never fell in love with Chrome/Chromium, because of Vimium, there was no command line, and often it happened, that key strokes has not been detected, at the time I played with it.

**Comment by marza91** [comment]
If I'm not mistaken, we're not allowed to change the GUI in any way using WebExtensions, right? It's a real downer if I can't have my gui="none", but I guess I/someone could fix that by hacking a bit at the firefox browser itself(?)

Regarding the current status line, would an alternative be to inject a fixed div/element into each open tab and have it stick to the bottom of the screen? Made a quick mock-up here to illustrate: https://marza91.github.io/vimperatortesting/

It would mean that there's a separate status-line being injected into each open tab/page when it loads, and I guess it could potentially mess with the styling of some pages(?), but if there's no alternative, I think it would be better than nothing?

You could (maybe?) do this with command line as well, if we put an input-element in the div, but I'm not sure if there are better ways of doing this in WebExtensions and/or if it would work. I'm going to try to learn some more WebExtensions stuff if I can find the time, and maybe make a statusbar/command mode prototype.

I could try to put this into the QuantumVim project (@shinglyu) as well, but for now, I can't really commit too much time into this. If no one have the time right now to take on a "project lead" role in the WebExtension port, maybe it's just as well if everyone tries to make the part they want the most themselves, and then we can look at merging stuff together later? It would require some coordination here or someplace else though, so we don't end up with 15 implementations of the same thing.

I'm not an expert at any of this stuff, just throwing out some ideas here, feel free to shut them all down :)

**Comment by mkoskar** [comment]
@marza91 Have a look at `cvim` Chrome extension. I does something what you are describing pretty well actually. I suppose Vimperator could take inspiration there, although it's pretty obvious since it seems like the only viable option anyway.

**Comment by protrolium** [comment]
The switch to webExtensions (and depreciation of XUL) in November is upsetting, since I am a user who thinks status bar/command-line + `set: gui=none` IS ESSENTIAL. In addition to everything @SammysHP mentioned, I also use `[[` , `]]` very often, as well as my own qmarks and `a` (for bookmarking).

**Comment by fakedrake** [comment]
My 2 cents: I had some free time and checked out the chrome source for building the UI. It looks like at least there there is not much room for modification. The layout is hardcoded in C++. Here are some interesting positions in the code:

- [Getting the tab bar visibility](https://cs.chromium.org/chromium/src/chrome/browser/ui/views/frame/browser_view.cc?type=cs&l=527)
- [Deciding what widgets should be shown](https://cs.chromium.org/chromium/src/chrome/browser/ui/browser.cc?type=cs&q=SupportsWindowFeatureImpl&l=2502)

Also I was thinking that for the non-UI stuff that we can't currently do without XPCOM (eg edit textfield in external editor) we could make a daemon in nodejs/python/go/language-du-jour and optionally interact with it over http to unlock that functionality.

**Comment by wcunning** [comment]
It seems to me that the right way forward is to get some api requests in as soon as possible. Given that thread already some discussion of the required keyboard shortcuts for VimFX, and I would bet on one for UI requirements for TreeStyleTabs, it would make sense to extend those requests or contribute to those apis to get what this project needs.

https://bugzilla.mozilla.org/show_bug.cgi?id=1215061

**Comment by fakedrake** [comment]
Are you aware of any existing similar thread addressing the UI (a quick
search didn't yield anything)

**Comment by wcunning** [comment]
https://wiki.mozilla.org/WebExtensions/Future

This sorta hints that they know it's an issue, but I'm not finding much either. I assumed that there would be something after reading about the VimFX API request. Still, I think that it shows at least a theoretical willingness to expose an API for those features, since Vimperator and TreeStyleTabs both need something in the same vein and are popular among the userbase.

**Comment by sirn** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=1215064

There's this one about implementing a custom toolbar. Nothing about hiding the toolbar, though.

**Comment by robehickman** [comment]
To me the current UI modification is absolutely essential to the value of this add-on. I've looked at similar things done in Chrome and found them unreliable and unusable.

**Comment by nachoparker** [comment]
Hello everyone,

I have been reading this and other threads and following the whole situation for a while. This is indeed worrysome. I do not want to create a wall of text, so I will try to be brief

- Mozilla is focusing the non-power user, and for that and other reasons, it has given up having powerful extensions that can do anything. This makes sense to try to recover user share in the market but leaves power users in a bad spot. It has conceded defeat.
- Web extensions or XUL it does not matter, but what you can do with it.
- There are many projects that need to reimplement the same thing to an extent
- Breaking compatibility is a high prize to pay, but FF internal architecture needs a rework to be more performant. e10s, servo and rust look really good and I think FF will come back really strong at the price of addons. Painful but worth it.
- A browser is a huge thing and we need to work over existing open source. It is just impossible to build a full fledged browser just to have vim-like navigation. Nowadays, browsers are as complex as operating system, and creating from zero is stupid. Many young developers like starting from zero and reinventing the wheel but it will go nowhere.

Given these facts, I think that the way to proceed is clear, but there is a need of manpower and cannot be done if we only argue about we need to have a command line or we need to remove the navigation bar.

- Get together (all projects) and create the basic minimum that every party agrees on. jklh navigation, follow hints... the most basic stuff. We should do this with web extensions to the extent firefox API allows us to.
- Once this is done, each project can fork and do their own particular taste, but we can share patches for the core.
- Then the ideal would be to FORK firefox. I am not talking about Pale Moon, which is a FF 27 and is not getting the latest things. We need the latest things. We need to take advantage of all the work that is being done that is open source and build on top of it. This is not a small project, and a browser needs to be performant and secure. We need to be smart. We should create a fork of firefox that allows power users to get what they deserve and install vimperator (or whatever) as an addon to that browser.
- On top of that power-user-firefox we can further develop the addon to have control of the UI and any other things that mozilla is trying to lock down in the name of safety for non power users.
- We should rebase new firefox upstream in a regular basis to keep the browser secure, performant , and up to date

If it were under this vision I would love to help. I could start analyzing the source code of FF, or maybe I could contribute to the addon itself.

I really think we need to be smart about this or we will get locked down to a Chrome clone that targets non technical users.

**Comment by The-Compiler** [comment]
> A browser is a huge thing and we need to work over existing open source. It is just impossible to build a full fledged browser just to have vim-like navigation. Nowadays, browsers are as complex as operating system, and creating from zero is stupid. Many young developers like starting from zero and reinventing the wheel but it will go nowhere.

I'll have to disagree. Building your own browser backend (rendering, JS, network stack, etc.) would be "stupid", but building a browser based on an existing engine such as [CEF](https://bitbucket.org/chromiumembedded/cef), [QtWebEngine](https://wiki.qt.io/QtWebEngine) (both based on Chromium) or [WebKit](https://webkit.org/) is very much possible. As a very simple example, take [suckless surf](http://surf.suckless.org/) which has basic vim-like browsing on top of WebKitGTK (an outdated version of it, sadly) and is [some 2000 lines of code](http://git.suckless.org/surf/tree/surf.c).

Of course, it is a lot of work, but I started such a project ([qutebrowser](https://qutebrowser.org/)) three years ago, and it's still going strong. It's exactly what you would imagine - a browser built on top of either Chromium (QtWebEngine) or WebKit (QtWebKit), with a minimal UI, and a lot of vim(perator)-like functionality. I've been using Vimperator and then dwb (another minimal browser project which died with WebKitGTK's massive API changes with WebKit2) for years, and then after not finding an alternative, started my own thing.

There's still always a lot of work to do (for example, it doesn't have something NoScript-like yet, but that'll hopefully come in the next few months), but I wouldn't say it's stupid. 
[No intentions of stopping either](https://github.com/qutebrowser/qutebrowser/graphs/contributors) :wink:. It has a solid day-to-day userbase by now - probably >1000, but it's hard to tell (it doesn't phone home). Other than other small browsers, it also works on Windows, OS X and others.

For larger things, crowdfunded full-time work is also a possibility. I did it very successfully (>7000€ raised, 2 months fulltime work on the new QtWebEngine backend based on Chromium) last year, and I'm already planning to do another one this year, to drive things forward as much as I can.

Foolish? Maybe. Stupid? Nope. Fun? Definitely (usually). :wink:

(That being said, of course I'd love Vimperator or something comparable for Firefox to live on - diversity is always a good thing.)

**Comment by robehickman** [comment]
I agree with 'The-Compiler' above, modern browsers are split into components. Most of the complexity is in the back-end engine which actually renders the HTML, handles JavaScript and what not. Different interfaces can be built on these engines fairly easily.

Nothing I've seen so far though has reached a level of being genuinely usable largely due to lacking extensions. The biggest one that is absolutely critical for me is LastPass. Coming close second uBlockOrigin. 

The web today is unusable without a blocker of some description. The NoScript approach is too heavy handed for todays world IMO.

**Comment by nachoparker** [comment]
I apologize, definitely the word 'stupid' was not appropriate.

I agree that it is really fun and interesting starting something from zero. I know your project, I have tried it (congratulations! :) )

That being said, I keep seeing fragmentation and people reinventing the wheel instead of joining forces.
 
Do you want to start your thing from zero, or from webkit? by all means! please, I do not want to deter anybody from doing that, let alone call them stupid (my apologies again).

But the case that matters for me right now and for many people is not to learn and have fun doing a browser (not only), but to have the power of FF and Vim together.

In my point of view it is a problem with open source that there is little manpower to compete with the big guys, for instance, Chrome, and before you know it you are locked with a propietary solution or free but managed by a big company ( I like to call that freetary ). Fragmentation is our weakness.

I do not want to reprogram what already has been done, I want to use the muscle behind Mozilla to keep FF my daily browser, but just change what I don't like, because that is what open source is about.

As a final note, if I wanted to use webkit or qtwebkit or whatever as a basis I would not be writing on a FF plugin github

All the best with your project, again, it is really cool

**Comment by nachoparker** [comment]
To further explain myself, I am getting inspired in this issue by the Xposed in Android.

I ignore if something like this would be possible in Firefox, but imagine we can find a way to inject code some of the basic "primitives" in FF in a standarized way.

Then, plugins could use this API (like Xposed android apps) and give back the power of FF to power users.

In this manner, we could get back noscript, vimperator, tab groups and all the things that we are going to lose really soon

**Comment by haasn** [comment]
> Mozilla is focusing the non-power user, and for that and other reasons, it has given up having powerful extensions that can do anything. This makes sense to try to recover user share in the market but leaves power users in a bad spot. It has conceded defeat.

Except this doesn't work in practice, because chrome users are already happy using chrome - not firefox. So trying to “regain” chrome users by turning your browser into a carbon copy of chrome does nothing except alienate the rest of your user base as well.

Instead of recovering market share, the only thing Mozilla has accomplished is killing off their market share even faster.

**Comment by nachoparker** [comment]
Please, I would like to keep the politics out of this. Mozilla's decisions are something we cannot control and complaining about them here will only bring noise to this thread.

**Comment by ohjames** [comment]
Firefox had more 500 contributors committing in the last 60 days. Chrome had almost double that. Forking it would be great if you could amass several hundred regular contributors, but you almost definitely can't. Palemoon couldn't. The fork would then inevitably end up eating dust. There's something to admire about foolhardiness but the case for pragmatism here is overwhelming.

**Comment by nachoparker** [comment]
I know it's not likely to happen... I guessed I could at least probe the crowd here.

Anyway, the idea is not to develop a new browser, but to closely follow the upstream and patch just enogh to get what we need, UI changes, noscript like capabilities, XPCOM like capabilities...

**Comment by lord-re** [comment]
Another way to explore could be starting from [servoshell](https://github.com/paulrouget/servoshell) ? or plain servo and a fork of [browser.html](https://github.com/browserhtml/browserhtml) ? Or something like [miserve](https://github.com/meh/miserve) ?
Going toward these will lost many firefox users using other addons but would lead to more flexibility and probably on a longer timescale be less dependant on mozilla.

**Comment by nachoparker** [comment]
Servo is Mozilla, also the best from Servo will be gradually integrated into FF. If we follow upstream we will eventually get big parts of Servo.

In my opinion we need manpower and that is why I suggest going for a popular browser with already good development and just change the minimum to meet our needs.

If we go too experimental I think the project will die as having a usable browser is a lot of work that needs hundreds of developers as it has been pointed out. Also FF has a nice userbase so we know that it is well tested, usable and safe.

**Comment by ohjames** [comment]
The best thing to do is port vimperator/pentadactyl or at least parts of them to WebExtensions API, compromising and removing features where necessary due to limitations. Each of those removals or changes would be documented and then the community would vote on which are most important. We could then use that evidence and the strength of the community to advocate for changes in the WebExtension API, building prototypes as patches against Firefox where necessary to strengthen the case (and also for us to use).

Any solution involving forking or custom browsers will just end be being a hobby project, at best there will enough interest to maintain it for years, but it won't have the longevity of Firefox or ever be a pragmatic choice for most developers, let alone mainstream users.

**Comment by The-Compiler** [comment]
> Anyway, the idea is not to develop a new browser, but to closely follow the upstream and patch just enogh to get what we need, UI changes, noscript like capabilities, XPCOM like capabilities...

As someone essentially writing a browser (see my post above), it's probably easier than forking Firefox.

I've more experience with this with Chromium, but I'd guess Firefox' situation won't be too different - there is a new Chromium major release [all 1-2 months](https://www.chromium.org/developers/calendar), and each release has millions of changed lines. In QtWebEngine, updating the subset (!) of Chromium they depend on from 53 to 55 meant over [5 million added and 3 million deleted lines](https://codereview.qt-project.org/#/c/181297/) (let that sink in for a second... 😆).

Keeping up with this (even with multiple active people, which is unfortunately rather rare) as a community project seems... difficult, to say the least. Qt says they need about a man-month for every Chromium update, and this time Chromium did some big changes in its build-system and it's been almost 2 months (with multiple full-time employees!).

**Comment by nachoparker** [comment]
@ohjames that is what I propose for the first step for sure... join forces

The forking thing would definitely be the last thing to do

**Comment by SammysHP** [comment]
Please let us not talk about forking Firefox or developing an own browser. This is a completely different project (and I really doubt that we will find enough developers to create a modern browser that is feature complete compared to Firefox or Chrome).

**Comment by bovine3dom** [comment]
@cmcaine and I have sketched out how Vimperator might be ported to WebExtensions and @cmcaine has talked on IRC with andym, a Firefox developer, about the more contentious bits.

The short of it is that we think a WebExtension Vimperator can be made to work a bit better than a cVim-port in Firefox, if certain APIs that the devs have accepted on principle are implemented and if the theming API allows the hiding of basically all the standard UI elements.

The good part is that cVim is really quite nice, and we have a lot of freedom to do our own thing.

The problems are these, in our order of priority:

1. The new theming API will not permit UI elements to be hidden, nor are any other APIs planned that would do that.
2. Some features on restricted pages (about:*, addons.mozilla.org) will not work, hinting and autocompletion on the commandline are the most important features we'll lose (we'll also lose Visual, Insert and Caret). We should be able to keep tab, page and history navigation and searching (if a keyboard API is implemented by FF devs: https://bugzilla.mozilla.org/show_bug.cgi?id=1215061; UI might be available with https://bugzilla.mozilla.org/show_bug.cgi?id=1215064).
3. We won't be able to access the filesystem as easily for e.g. plugins, unless a filesystem API is accepted and implemented. There is already a storage API which can synchronise with Firefox Sync: https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/storage
4. Won't be able to run external processes.

We consider 4 and 3 to be unimportant issues; 3 because it can be worked around and 4 because we never use it.

The only workaround for 1 that we have so far is if the theming API is flexible enough that we can implement auto-hiding tree style tabs or otherwise hide Firefox UI, but that seems unlikely.

We are hoping to get a productive discussion going on whether it will be possible to use content scripts on all pages here:
https://bugzilla.mozilla.org/show_bug.cgi?id=1342379

The theming API is being tracked here, but the developers say that they aren't ready to discuss it yet:
https://bugzilla.mozilla.org/show_bug.cgi?id=1330328

I think it is important that we engage productively with the Firefox developers on this. We should be able to reach a reasonable compromise.

TL;DR: We can already do most of the stuff we need/want to in WebExtensions, with the exception of hiding Firefox's UI, and getting trapped on special pages (e.g about:config).

**Comment by The-Compiler** [comment]
> and 4 because we never use it.

What about `Ctrl-i` to spawn an external editor (i.e. vim for example) to edit textareas and other inputs?

**Comment by robehickman** [comment]
"What about Ctrl-i to spawn an external editor (i.e. vim for example) to edit textareas and other inputs?"

I use this all the time.

**Comment by dpaetzel** [comment]
> What about Ctrl-i to spawn an external editor (i.e. vim for example) to edit textareas and other inputs?

We could try to get something similar to surfingkeys' integrated vim-like editor (https://github.com/brookhong/Surfingkeys#vim-editor)?  Just a suggestion as I don't know much about firefox and/or chrome extensions…

**Comment by robehickman** [comment]
The value of vimperator to me is that it completely replaces the stock UI with one that's superior. The stock UI is now useless and consequently I don't want to see it.

**Comment by bovine3dom** [comment]
> What about `Ctrl-i`

I've never used it - if I'm writing something big enough that it warrants Vim, I generally write the whole thing in Vim and then copy/paste it into the text box, because I'd be concerned about losing it, and I want access to e.g, Undotree. I did that with my previous comment.

As @dpaetzel says, there do seem to be plenty of JavaScript vim-like editors (e.g, http://coolwanglu.github.io/vim.js/emterpreter/vim.html) that could fill that gap.

**Comment by lord-re** [comment]
Maybe splitting vimperator into sub-addons ?
One for messing with the UI (when the new api will permit it) : vimperator-ui
One for the keybindings : vimperator-keys
One for the external|vim-like editor : vimperator-editor

**Comment by nachoparker** [comment]
`ctrl+i` is very nice. I also use `ctrl+t` quite a bit.

Thanks to @cmcaine and @bovine3dom for taking the iniciative. Let me know how I can help. I can code.

**Comment by nagisa** [comment]
A somewhat important feature to me that hasn’t been mentioned in the list above is `VIMPERATOR_INIT` environment variable, which may be… difficult… to make work.

One of the most used `about:` pages is `about:downloads` for me, and it would be a shame if I had to resort to tabbing to navigate that page.

**Comment by SammysHP** [comment]
We should create a list of all people (from Vimperator, Pentadactyl...) who want to contribute to the new addon. We had maybe five contributors for Vimperator last year, not enough to make a new addon.

**Comment by klaplong** [comment]
Yes, please add me! I'm not experienced in writing addons but I do have a lot of programming and JS experience. I actually have contributed 2 years ago. Don't have a load of time but am willing to invest some into this project!

**Comment by nachoparker** [comment]
alright, let me count:

nachoparker
klaplong
tfrangio
apheleia
svarogg
cmcaine
bovine3dom

Anyone else? 

There, better get moving or we'll be talking here forever. I am not aware of any repo that has been created for this (correct me if I am wrong) so created a repo

https://github.com/nachoparker/vimperator2

We can discuss over there how to start... from cvim codebase? vimperator codebase?

**Comment by mikejkjr** [comment]
If there's room, I'd love to help as well. I've used Vimperator for quite a while and would hate to see it die. I'm on vacation right now but after March 4 I'll have spare time to contribute.

I'm fairly experienced in JS development, but haven't done any plugin work before.

**Comment by Segaja** [comment]
I would try to help aswell. And maybe it makes sense to make this repository in the vimperator group?

**Comment by tecfu** [comment]
@gkatsev @maxauthority @SammysHP @timss 

I suggest that a Vimperator WebExtension port should be created under the custody of the current github account used for Vimperator. 

- Will be easier to find
- Leverages access to the existing user community
- Encourages leadership and participation from existing contributors
- Reduces likelihood of fragmentation 

@nachoparker Therefore I'd like to suggest a different approach: ask the current maintainers to create a new vimpertator/vimperator-labs repository and then petition for push privileges (or to become an additional Vimperator maintainer).

**Comment by nachoparker** [comment]
@tecfu I am fine with that, as long as we get moving :)

updated list :

nachoparker
klaplong
tfrangio
apheleia
svarogg
cmcaine
bovine3dom
segaja
mikejkjr

**Comment by tecfu** [comment]
@nachoparker Cool. Let's see if we can get one of the current maintainers (@gkatsev @maxauthority @SammysHP @timss) to respond and facilitate.

**Comment by nachoparker** [comment]
sure, my apologies to the maintainers. I'd also rather do it under their leadership :)

**Comment by tecfu** [comment]
@nachoparker Don't sweat it dude, I don't think anyone sees your suggestion as being subversive - actually I think its the norm and the spirit of Github in general. But - in my observation projects that retain brand, contributor, and user continuity seem to outperform those that don't. 

I have a recent experience of this on my own trying to get buy-in on a fork I did of https://github.com/Automattic/cli-table. I think it can become a big competition for attention and recognition rather than a collaborative effort to improve and maintain a product. I just didn't want to see Vimperator go the same route.

**Comment by bovine3dom** [comment]
If people are motivated right now, it would be useful to compile a list of features that they feel are important to reimplement, because apparently there are still lots of them I don't know about :)

**Comment by nachoparker** [comment]
agree, like I said from the beginning, the more the merrier. I actually linked this conversation on the pentadactyl github. I think it would be ideal that we (pentadactyl, vimperator, vimfx and whatnot) created a solid base and then everyone can fork to their taste if they want different things.

**Comment by tecfu** [comment]
> I actually linked this conversation on the pentadactyl github.

@nachoparker Promoting a unified effort to create a Vim-like browser add on based on WebExtensions is a great idea since:

"To a large extent the system is compatible with the extension API supported by Google Chrome and Opera." - https://developer.mozilla.org/en-US/Add-ons/WebExtensions.

**Comment by SammysHP** [comment]
@tecfu I'd like to work together with the developers of other similar addons like Pentadactyl. Thus a repository in https://github.com/vimperatur might not be the best idea. ;)

**I created a chatroom on Gitter where we can organize and talk about the next steps: https://gitter.im/vim-webextension/Talk**

**Comment by nachoparker** [comment]
@tecfu thanks for linking all these projects here

For everyone that is new to this conversation, 

- the first goal is to have a common ground for "vim implementation in web extensions" that we can all contribute to and use.
- after this, we can all go back to having our own fork if we have different points of view, but we can still reuse the core

**Comment by cmcaine** [comment]
@nachoparker, @bovine3dom and I already have a repo with some relevant notes in it. I've renamed it and removed some unrelated stuff. It's available here: https://github.com/cmcaine/tridactyl

Almost all notes rather than live code, because we're waiting to find out what APIs and permissions the Firefox developers are willing to support.

The most interesting documents will be design.md and bug-message.md, probably.

**Comment by philc** [comment]
Hey folks, note that we intend to release Vimium for Firefox in the near term. You can track the work [here](https://github.com/philc/vimium/issues/2425).

**Comment by tecfu** [comment]
So far I've discovered that there are two codebases that already support a Vim-like add on using the WebExtensions API:

- https://github.com/shinglyu/QuantumVim
- https://github.com/mrmr1993/vimium/tree/firefox-merger

I'll also make mention to a repo which currently has only notes but which I will watch for development:

- https://github.com/mrmr1993/vimium/tree/firefox-merger

For now I'm going to watch these projects and revisit this issue in a month or so to reassess progress then.

**Comment by nachoparker** [comment]
I aggree, does not seem that we need to create yet another thing. Those look good

**Comment by cmcaine** [comment]
The existing chrome vim addons should work with Firefox with minimal changes.

My main interest in the tridactyl repo and in my bug reports on bugzilla is in encouraging the FF devs to allow a better UX than is currently possible with the chrome WebExtension APIs. There's definitely interest in enabling this from Firefox developers.

There is a need for developer interest in a keyboard API proposal for Firefox: the developers are OK with maintaining it but the initial implementation and definition was started by lydell (vimfx), but he's given up on it for now.

See: https://github.com/lydell/webextension-keyboard https://bugzilla.mozilla.org/show_bug.cgi?id=1215061

If you're interested in working on the keyboard API maybe get in contact with me and we can collaborate (email or issue on cmcaine/tridactyl).

@bovine3dom and I don't think there's much to be gained by outright forking an existing project, though we will certainly be reading the existing projects and adapting code. We will be experimenting with potential architectures this weekend and continuing to talk to the Firefox developers. If you're interested in working with us, please do get in contact, as above.

We'll post progress updates to this thread occasionally as @bovine3dom did today (unless anyone asks us to stop!).

Finally, it's worth noting that although Vimium is very good at what it does, it's a lot more lightweight than Vimperator: it's more like e.g. VimFx. I say this just to manage the expectations of some in this thread :) A closer WebExtension relation to Vimperator would be something like cVim or vrome.

**Comment by robehickman** [comment]
Just tried CVim on chromium. It has the same unreliability issue I've experienced with these add ons generally. There are instances where pressing a command key does nothing for no apparent reason. Clicking on the page tends to fix it but that also defeats the point of the extension.

**Comment by dpaetzel** [comment]
> There are instances where pressing a command key does nothing for no apparent reason. Clicking on the page tends to fix it but that also defeats the point of the extension.

While I do not find the source for that, I remember reading something about the reason for such a behavior being that the whole keybinding stuff for these chrome extensions is (always, irc) implemented as a user script that gets injected into every page and sometimes there seem to occur problems with the asynchronous loading of these scripts...

**Comment by robehickman** [comment]
> While I do not find the source for that, I remember reading something about the reason for such a behavior being that the whole keybinding stuff for these chrome extensions is (always, irc) implemented as a user script that gets injected into every page and sometimes there seem to occur problems with the asynchronous loading of these scripts...

Whatever causes it, it needs resolving if Firefox is going to have anything resembling Vimperator. I have never had a problem with it not responding.

**Comment by cmcaine** [comment]
@robehickman I didn't experience that. Are you able to identify any correlated circumstances?

As @dpaetzel says, if it happened shortly after the page was loaded it could be because the content script hasn't been loaded yet.

**Comment by Segaja** [comment]
It would be resolved if Firefox implements a normal keyboard API so that the addons don't have to rely on such hacky workarounds.

**Comment by robehickman** [comment]
@cmcaine Not as such, it just stops randomly. For example flicking through tabs with 'gt', after a few successful changes it stops responding.

**Comment by cmcaine** [comment]
@robehickman Thanks. More reason to get the keyboard API in Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1215061

**Comment by robehickman** [comment]
@cmcaine Presumably it is related to script loading as refreshing the page fixed it. Agree having a better way to do this would be preferable.

I'm not certain exactly how the input system works but any keyboard API for add ons needs to be able to interrupt the event flow like vimperator currently does. It catches all events unless pass-through is enabled.

**Comment by cmcaine** [comment]
By interrupt the event flow, do you mean stopping the key getting through to the browser UI or the page?

We can definitely do that, even with content scripts (though can't stop all browser shortcuts with content script listener).

**Comment by robehickman** [comment]
> By interrupt the event flow, do you mean stopping the key getting through to the browser UI or the page?

That's what I meant.

**Comment by maxauthority** [comment]
> Hey folks, note that we intend to release Vimium for Firefox in the near term. You can track the work here.

@pihlc: Awesome! I always loved Vimium on Chrome (the only exception being 'o' not reusing/focussing the (really good) Chrome-omnibar to open urls, but that's a WebExtensions limitation). Anyway, a proper Vim(perator)-like extension which works on all major browsers (maybe soon Edge as well) is certainly great. And while we can whine about the loss of the power of XUL-based add-ons, we just cannot change that, and should embrace the future with a high-quality vim-like add-on, be it Vimium and/or a Vimperator based WebExtension (but even the, using Vimium's source as a basis would be a tremendous help).

BTW: I added a reference to Vimium a few days ago to the vimperator.org front page, as I really think promoting good vim-like add-ons for any host application is also in the spirit of Vimperator (users and developers).

**Comment by bovine3dom** [comment]
I forgot to post this here weeks ago:

Since the version on the AMO is still not working with FF51.0.1, here's a packaged one that @cmcaine signed, and currently works.

[vimperator-3.16.0.20170209194233-signed.zip](https://github.com/vimperator/vimperator-labs/files/802168/vimperator-3.16.0.20170209194233-signed.zip)

If you want to be sure it isn't full of viruses, etc, build it yourself - but hopefully it will help someone.

**Comment by timss** [comment]
3.16 is on AMO now though?

https://addons.mozilla.org/en-Us/firefox/addon/vimperator/versions/?page=1#version-3.16.0

**Comment by KamphuisAlex** [comment]
I'm very willing to throw some hours at this. I'll see it as my way of saying thank you for all the time Vimperator has saved me so far.

@nachoparker are you willing to take on a team member that needs to learn a lot on-the-go? Recently I've only done Python and C++

**Comment by bovine3dom** [comment]
@KamphuisAlex I think the most useful thing you (or anyone else interested in helping) could do at the moment would be to start making a list of features that you care about the most (and least). That will give us a better idea of what new WebExtensions APIs we need to lobby the Firefox developers for. I've started a list in #729.

**Comment by nachoparker** [comment]
I aggree with bovine3dom, let's start with the list.

> @nachoparker are you willing to take on a team member that needs to learn a lot on-the-go? Recently I've only done Python and C++

Any help will be appreciated! I do not know who will lead this, but it won't be me. Probably current maintainers, or maybe @bovine3dom and @cmcaine. First I do not have as much time as I would like to, second, I have developed XUL addons long time ago, but I am rusty. Third, they were already on this and have done good research that we should use.

I will humbly contribute, I just wanted to get the thing to start moving ;)

**Comment by mrmr1993** [comment]
Update on Vimium FF compatability:
* I have a PR on Vimium philc/vimium#2476 which fixes the rest of the easiest issues.
* my branch [here](https://github.com/mrmr1993/vimium/tree/hud-focus) is the current state of the art, including an extra fix that makes searching work less badly.

The biggest improvements we could make would be by getting some Firefox issues fixed. Fixing the following (in order of priority) would almost completely close the gap with Chrome:
* *Chrome compatability - WebExtensions should be able to create "real" mouse events* - [issue 1356309](https://bugzilla.mozilla.org/show_bug.cgi?id=1356309)
* `window.find` throws sometimes, especially when wrapping (no issue, I'm scared to poke at it in case it gets removed/more broken)
* *Implement a clipboard api that supports copying arbitrary strings to the clipboard* - [issue 1344410](https://bugzilla.mozilla.org/show_bug.cgi?id=1344410)
* *WebExtensions: chrome.tabs.create fails if url is about:newtab* - [issue 1275209](https://bugzilla.mozilla.org/show_bug.cgi?id=1275209)

It would be a great help if people would vote for these (or otherwise persuade the FF devs to work on them).

**Comment by cmcaine** [comment]
Interesting work!

I've been working on a low level keyboard API to be added to Firefox on the basis 1) keyboard input isn't currently available on restricted pages without kludge and 2) that users report unreliable or otherwise problematic keyboard input on projects like Vimium in chrome.

I've not been able to reproduce (2). Have you encountered it?

**Comment by mrmr1993** [comment]
>  2) that users report unreliable or otherwise problematic keyboard input on projects like Vimium in chrome.
> I've not been able to reproduce (2). Have you encountered it?

Yes, but only within the scope of philc/vimium#2432. I'll have a look at that tomorrow, work permitting.

**Comment by maxauthority** [comment]
Matthew: Thanks for your update (and work!) on Firefox compatibility, I
think Vimium will become the first cross-platform vimperator-like browser
extension!

Is there a test version to quickly test (for curiousity) or do I need to
build my own XPI? Actually I stumbled upon
https://addons.mozilla.org/en-US/firefox/addon/vimium/ which certainly
isn't your addon, given it's age of 5 years.
In the medium-term one might think to talk to the Mozilla guys to see about
the name-clash, especially since I don't expect the "other" Vimium to ever
get updated to WebExtension-only firefox.

**Comment by mrmr1993** [comment]
> Is there a test version to quickly test (for curiousity) or do I need to build my own XPI?

I'm out for the rest of this evening, but I can post my latest XPI tomorrow. Otherwise, building Vimium using the normal instructions (current master has Firefox support now) will work.

> In the medium-term one might think to talk to the Mozilla guys to see about the name-clash...

cc-ing @smblott-github and @philc on this

**Comment by mrmr1993** [comment]
> Is there a test version to quickly test (for curiousity)

https://github.com/mrmr1993/vimium/releases/download/1.59.4/vimium-1.59.6-an.fx-linux.xpi

That should link to a build of the latest master.

**Comment by smblott-github** [comment]
Folks,

We have an experimental port of Vimium to Firefox here: [https://addons.mozilla.org/en-GB/firefox/addon/vimium-ff/](https://addons.mozilla.org/en-GB/firefox/addon/vimium-ff/).

Some Vimperator users might be interested in trying it out.

This port remains experimental, feedback (and PRs) would be most welcome over in the [Vimium project](https://github.com/philc/vimium/issues?q=is%3Aopen+sort%3Aupdated-desc).

**Comment by maxauthority** [comment]
Thanks a lot! :)

I tested the old version a few weeks ago and while (understandably for a
beta version) lacks a few features/bugs, it's still great :) Kudos to you
folks.

Ah one small thing, the description on the AMo link writes "The is a port
of the popular Chrome extension to Firefox." which doesn't sound like
proper English ;)

**Comment by ohjames** [comment]
Firefox has become so slow (especially with developer tools open) that the only reason I'm even using it is because of vimperator. From my perspective spending time porting a chrome extension over to firefox is wasted, if I have to give up vimperator I'll be giving up firefox also, and I wonder if many other people here are in the same position.

I want to support open source alternatives but firefox is so substantially worse than chrome now that I won't be seriously inconveniencing myself to do it.

**Comment by tidux** [comment]
Firefox is slow for you BECAUSE of your addons.  A big reason for the webextensions restriction is that a lot of XUL addons don't work well or at all with multiprocess.  Switching to multiprocess-aware addons makes Firefox faster than Chrome in my experience, especially since FF54 when Linux finally got full OpenGL accel.

**Comment by nagisa** [comment]
I'd rather have a slow browser that has tolerable UX than a fast one with a
shite UX personally, but overall a fair point.

**Comment by GuillaumeLeclerc** [comment]
I see a lot of discussions about other plugins. Does that mean that Vimperator will die ? This is also the only reason I use Firefox.

**Comment by SammysHP** [comment]
Probably. The old code won't work anyway, so it would be a new start from scratch. That's why the focus is on other extensions at the moment.

**Comment by bovine3dom** [comment]
Yes.The whole thing would have to be rewritten from scratch; Vimperator in its current form will die.@cmcaine is intermittently trying to make the most Vimperator-like plugin he can. The main blocking issue (other than motivation) is the lack of APIs in new Firefox, and the fact that everything has to be written in JavaScript, a dreadful web-first language that will never catch on.Vimium/Vimfx are promising but have slightly more humble goals.                                                                                                                                                                                                                                                                                                                                                    From: notifications@github.comSent: 22 June 2017 19:33To: vimperator-labs@noreply.github.comReply to: reply@reply.github.comCc: oliver.blanthorn@postgrad.manchester.ac.uk; mention@noreply.github.comSubject: Re: [vimperator/vimperator-labs] Port to WebExtension (#705)  I see a lot of discussions about other plugins. Does that mean that Vimperator will die ? This is also the only reason I use Firefox.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by ohjames** [comment]
> a dreadful web-first language that will never catch on

Lol what JavaScript is the most popular language in the London contract market right now. It already caught on. I guess you don't code for money :P

**Comment by Shulyaka** [comment]
> Lol what JavaScript is the most popular language in the London contract market right now. It already caught on. I guess you don't code for money :P

I'm glad you have no objection of it being dreadful and web-first.

**Comment by ohjames** [comment]
You probably have zero knowledge of JavaScript beyond ES5.1, it's become a fairly nice language.

**Comment by timss** [comment]
This issue isn't for discussing the JavaScript programming language.

**Comment by cmcaine** [comment]
@bovine3dom was being sarcastic.

> JavaScript, a dreadful web-first language *that will never catch on*.

Emphasised bit was a hint :)

I'll have another crack soon at the firefox keyboard API I was working on.

So far we've come up with a broad architecture for our vimperator clone, learnt the development loop for webextensions, how to use the existing APIs, how to develop our own webextension APIs and tried a few javascript variants (probably going to use typescript).

Collaborators are still welcome to email me. I prefer working with a group, so even if you don't contribute much work of your own, you would catalyse me :)

**Comment by The-Compiler** [comment]
I've just updated the [list of similar projects](https://github.com/qutebrowser/qutebrowser#similar-projects) for [qutebrowser](https://www.qutebrowser.org/) - I added [Vimium-FF](https://addons.mozilla.org/en-GB/firefox/addon/vimium-ff/) and [Saka Key](https://github.com/lusakasa/saka-key) as Firefox extensions which are still active and using WebExtensions. Is there any project I've missed?

**Comment by cmcaine** [comment]
@The-Compiler I hadn't heard of saka key, thanks for the link :)

@kmaglione of Mozilla has agreed to do an initial review of my [keyboard api](https://github.com/cmcaine/keyboard-api) proposal a little while ago. The API isn't complete, but the skeleton is there and I wanted someone from mozilla to quickly check it over.

He and others at mozilla have said that the API is something they want in mainline firefox, but it's not a priority for them. It's yet to be seen if my implementation is ok or not.

In the meantime, development of tridactyl is slow because if the APIs aren't available, we won't be able to make a nice enough product (and we'll use qutebrowser instead, probably :)). We (mostly @bovine3dom) have ported our existing tech-demo style code to typescript (from literate coffeescript) and Olie's written an initial normal-mode parser.

The design is fairly simple and can probably be put together quite fast. We just need the required APIs.

If anyone would like to help, the next most important API (that has Mozilla preliminary approval) is a "keyboard shortcut/navigation" API that would allow a webextension to easily perform the exact actions that are possible with keyboard presses in firefox programmatically. This API is important to allow vim-style control of the restricted pages.

@lydell wrote about it here: https://github.com/lydell/webextension-keyboard/#user-content-browserkeyboardshortcutactions

If you're interested in developing or helping to develop that, you could make all of these vimperator-alikes better :) If you want some pointers, [email me](https://xkcd.com/138/).

**Comment by wnereiz** [comment]
I tried Vimium-FF and Saka Key. It is not working well for me as expect. For example, the tab switching does not work when the web page are not loaded yet. So they are not comparable with vimperator/VimFx.

I think the final solution is to add the APIs to support keybindings - like what @cmcaine is doing. I hope Mozilla developers merge the code asap.

**Comment by cmcaine** [comment]
The keyboard API I'm developing and the new [keyboard shortcut API](https://github.com/Koushien/keyboard-shortcut-api) that Koushien is developing will not land in FF57.

I've been asking for my work to be reviewed by a Mozillian for quite a while, but they're understandably busy with the push for FF57.

@bovine3dom and I think the best option for now is to develop or nominate some successor project and hope the APIs hit in 58. Last time we talked about it, Olie and I agreed to develop *something* for FF57 in order to build marketshare (why we care about that, idk :). We're still developing our own because, with the exception of cVim, none of the other projects have a proper ex-mode/statusline (and the cVim code isn't documented well).

Progress is slow, but we'll try to pick it up and deliver something more or less as good as cVim before the FF57 release (2017-11-14). Anyone who wants to help should mail me 👍

**Comment by R030t1** [comment]
> Mozilla's decisions are something we cannot control and complaining about them here will only bring noise to this thread.

While I don't think only complaints are a good idea, acting like the people here have no input will only prevent the Firefox developers from ever learning that their proposed changes are counterproductive to increasing the usability of their browser. If anyone has the time available to inquire about the design choices on the appropriate Mozilla mailing lists, I think doing so would be a good idea.

**Comment by nonameentername** [comment]
I like the idea of using webkit to build a vimperator browser https://github.com/qutebrowser/qutebrowser

**Comment by robehickman** [comment]
The problem with building something custom is that you loose all the existing addons. For me UBlock Origin and LastPass are utterly essential and I won't use any browser that does not support them.

**Comment by jvican** [comment]
@robehickman Note that you can use Bitwarden in FF57 instead of Lastpass. Ublock Origin is already supported.

**Comment by robehickman** [comment]
@jvican My comment is directed at @nonameentername sugestion of building a custom browser using webkit. I am aware that other addons are supported and I'm not interested in changing my password manager. LastPass got there Crypto right and the addon has been battle tested, and improved from security researchers attacking it.

For me the killer feature of Vimperator is it's ability to dispense with the stock UI. Once you have good keyboard control the only thing the UI does is waste screen space.

**Comment by Allaman** [comment]
I have read through the whole discussion but now I am not more clever than before. Can someone please point me to a roadmap or any plans regarding vimperator? For me, as for many others I think, vimperator is crucial and the last reason to stick to Firefox. I am a heavy power user for years now and an end of vimperator would greatly impact my browser workflow as there are no real equal alternatives in my opinion.

**Comment by Hi-Angel** [comment]
@Allaman if you don't have webextension-addons essential for you, [try qutebrowser](https://github.com/qutebrowser/qutebrowser/). It's basically a frontend to Chromium's engine *(prior to 1.0 version needs to be started with `--backend webengine`)*, which is pretty much like Vimperator or Pentadactyl. *(must be noted, there is [a plan in the future to support API for addons](https://github.com/qutebrowser/qutebrowser/issues/30), might as well be webextensions)*.

I've migrated last weekend to qutebrowser, configured a little bit to my likes — and so far my experience goes very well. It's even better than Firefox+Pentadactyl was — e.g. Pentadactyl upon starting editor on many sites often used to insert a xml junk, and to leave the site screwed even if you immediately closed the editor. But with qutebrowser it works just fine.

**Comment by anhtuann** [comment]
@Allaman I also moved to qutebrowser since last Firefox update (v.56) which broke vimperator for me. And I am really satisfied.

**Comment by Allaman** [comment]
@anhtuann yes, FF 56 broke it for me too ...
@Hi-Angel I actually tried quotebrowser a while ago and it looked very promising but my firefox **was** just to superior.  `...configured a little bit to my likes` would you like to share your config? :)

**Comment by Hi-Angel** [comment]
@Allaman yeah, [here it is](https://gist.github.com/Hi-Angel/39ed51b752fb28a9d582fd675d83ab3a). You might want to look with diff to original one in case you wouldn't like something. However it should be noted that I am using qutebrowser-0.11.0, whilst the soon to be released 1.0 version have significanly changed configuration, so the old config files won't work there.

> I actually tried quotebrowser a while ago and it looked very promising but my firefox was just to superior

I think you've tried it with webkit engine. It was default, and is indeed lags a bit. I'm using QtWebEngine backend instead *(e.g. `--backend webengine` option for 0.11, and default in 1.0)*, and it is significantly faster than Firefox+Pentadactyl ever was. I mean, really, e.g. site Phoronix have been loading 2-3 seconds for me with Firefox. I always thought it's the site slow — but guess what, qutebrowser loads it instantly! D: It might be related to its adblock though — not sure. Btw, use `adblock-update`, and may be take a look at look at [this hack to disable youtube ads](https://github.com/airodactyl/qutebrowser/commit/48eb50d7e6160a2439258a62fbe92e9623e61124) *(qutebrowser is using python, so you can directly edit the file in the package)*.

**Comment by Hi-Angel** [comment]
It should be noted though that I disabled adblock there through `host-blocking-whitelist = *phoronix.com/*,lwn.net/*` as you can see in the config. But, well, certain sites behave oddly, and I can't say if it works or not *(e.g. perhaps shall I use whitespace as a separator? I copied made it through comma because it's the way it works in other places)*.

**Comment by cmcaine** [comment]
If you don't need new FF features, Firefox ESR is compatible with vimperator and pentadactyl and will be until May next year, I think.

There's a fork of Firefox called waterfox which is also vimperator compatible but they're going to struggle to backport changes as Firefox diverges.

If you require compatibility with WebExtensions or latest Firefox/chrome the closest currently available is cVim for chrome. In the near future we hope Tridactyl will also fill this space, but for Firefox.

If you're ok with a lightweight alternative vimium and saka-key are both good.

I'm working with Koushien to get new APIs approved for Firefox that will address some of the shortcomings with WebExtensions, but these won't land until FF58 at the earliest.

If you don't need add-ons Qutebrowser is the best of the standalones I've tried. But there are others you might want to give a go. Qutebrowser readme contains an up-to-date list.

Last time I tried qutebrowser the adblocker wasn't very good, but it may have improved.

**Comment by vindarel** [comment]
Hey, just discovered [Vim Vixen](https://github.com/ueokande/vim-vixen) and it's fantastic so far because its keybindings are very similar to Vimperator's: t, T, b, o, O, y, Y, d, u, b, search engines completion… a prompt with `:` (limited, no completion). Again it does not work on special pages and sometimes it gets disabled for no reason (or, just after a call to the prompt). I prefer it to Vimium-fx so far.

**Comment by iago-lito** [comment]
I also plan not to upgrade to FF57 as long as no vimperator/pentadactyl/any-vim-like navigation has been ported to it. But I understand it seems not easy.

I have no idea how this would translate into actual *collaboration*, but since:

- FF needs and does undergo an aggressive upgrade (which is *legitimate*)
- FF changes its philosophy about power user (which is *legitimate*)
- FF new release breaks Vim/Penta/*+.∞ (which is a *fact*)
- Vim/Penta/*+.∞ will be hard to port under new FF philosophy (which is a *fact*)
- Vim/Penta/*+.∞ users are numerous and love this navigation style (which is *legitimate*)
- They love it so much that they are reluctant to upgrade (which is.. a *weird* situation)
- Vim/Penta/*+.∞ is.. intimely tied to Firefox software (so it'd feel wrong to *spin off*)

Is there not any way to make *Firefox* itself our main negotiating partner?  
Is there not any line they can answer and listen to our community?   
Would our navigation style and our scope statement be difficult for them to implement without breaking their new objectives?  
How much help would they need?  
How much *money* would they need?
Would they not be interested in selling an option like _**« Firefox special unique build-in superfast comfort rust-powered modern coder-like original expert remarkable exotic mango-flavoured alien incredible pure keyboard navigation feature for real efficiency lovers »**_ or anything sugar-punchy like that, which would even enlarge our community?

---

I a nutshell: Does Firefox even know what is happening here? Would she give us a hand?

**Comment by tidux** [comment]
Vim Vixen works fine on Nightly.

On November 14, 2017 10:05:18 AM PST, Iago-lito <notifications@github.com> wrote:
>I also plan not to upgrade to FF57 as long as no
>vimperator/pentadactyl/any-vim-like navigation has been ported to it.
>But I understand it seems not easy.
>
>I have no idea how this would translate into actual *collaboration*,
>but since:
>
>- FF needs and does undergo an aggressive upgrade (which is
>*legitimate*)
>- FF changes its philosophy about power user (which is *legitimate*)
>- FF new release breaks Vim/Penta/*+.∞ (which is a *fact*)
>- Vim/Penta/*+.∞ will be hard to port under new FF philosophy (which is
>a *fact*)
>- Vim/Penta/*+.∞ users are numerous and love this navigation style
>(which is *legitimate*)
>- They love it so much that they are reluctant to upgrade (which is.. a
>*weird* situation)
>- Vim/Penta/*+.∞ is.. intimely tied to Firefox software (so it'd feel
>wrong to *spin off*)
>
>Is there not any way to make *Firefox* itself our main negotiating
>partner?  
>Is there not any line they can answer and listen to our community?   
>Would our navigation style and our scope statement be difficult for
>them to implement without breaking their new objectives?  
>How much help would they need?  
>How much *money* would they need?
>Would they not be interested in selling an option like _**« Firefox
>special unique build-in superfast comfort rust-powered modern
>coder-like original expert remarkable exotic mango-flavoured alien
>incredible pure keyboard navigation feature for real efficiency lovers
>»**_ or anything sugar-punchy like that, which would even enlarge our
>community?
>
>---
>
>I a nutshell: Does Firefox even know what is happening here? Would she
>give us a hand?
>
>-- 
>You are receiving this because you commented.
>Reply to this email directly or view it on GitHub:
>https://github.com/vimperator/vimperator-labs/issues/705#issuecomment-344345461

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by haasn** [comment]
> Is there not any way to make Firefox itself our main negotiating partner?

Firefox doesn't decide anything. The right negotiation partner is Chrome. Firefox obviously just copies whatever chrome does 1:1

**Comment by cmcaine** [comment]
> Firefox doesn't decide anything. The right negotiation partner is Chrome. Mozilla obviously just copies whatever Google does 1:1

That is not true. The WebExtension APIs offered by Firefox already contain many features that chrome lacks and I am working constructively with them to add the APIs we need to make something a lot like vimperator in Firefox.

Tridactyl only implements a small fraction of vimperator features now, but we looked at the feasibility of almost all vimperator features before we began in earnest.

At present, there is little functionality from vimperator that we cannot replicate in the WebExtensions API now and we will be able to do most of the rest with new planned APIs or extensions to the existing APIs.

We can't read or write the values on about:config, nor can we change the UI in the same way (though we have a very good userChrome.css). We are also very limited on some restricted pages (about:*, file:*, etc), but I and others have persuaded Mozilla to lift the most onerous restrictions soon and this situation will continue to improve.

**Comment by bovine3dom** [comment]
Just to add to what cmcaine said: the main reason Tridactyl is lacking in features is because of a lack of manpower and time, not because of fundamental issues with WebExtensions.

**Comment by iago-lito** [comment]
@cmcaine Great! I am glad to find out about [Tridactyl](https://github.com/cmcaine/tridactyl). I think you are doing things the right way; communicating with Mozilla API devs, and having first checked you could replicate everything.  
Thank y♡u for your involvement and your patience, guys, and keep up with the good work :)

**Comment by robehickman** [comment]
Firefox updated itself and have lost vimperator. Vim Vixen is unuseable because. it dosn't support gt/gT to switch between tabs and does not work in some interfaces including the firefox addons website. At least it supports 'Y' to copy the current URL. Any other options?

'gt' and loaded in ALL tabs is essential for how I use this as I frequently flip through tabs with those keys and as I've been using that every day for going on 8 or 9 years it is deeply ingrained in my muscle memory.

I am not at all impressed by the lack of UI customization in this Firefox update. The back/forward buttons in the UI, the URL bar and other stuff are things that I literally never use as I either have shortcuts for them or can show the same information contextually. The current page URL is useless information and thus a waste of screen space most of the time. If I actually need to see it I can do so with shift-o.

**Comment by robehickman** [comment]
Addition: Vim-vixen has configuration options and it can be made to use 'gt' easily. ':q' to close current tab does not work and can't see any way to add that in the settings. Dosn't affect it being broken in some tabs.

**Comment by Koushien** [comment]
@robehickman 
> does not work in some interfaces including the firefox addons website

AMO (addons.mozilla.org) and about:* pages don't load webextension scripts for security so it isn't specific to Vim-vixen. The [Vimperator readme] lists several alternatives, but keep in mind these projects are still young and implementing features to accomodate such webextensions isn't top priority for Mozilla. If you don't have the leisure to experiment, I'd suggest sticking with Firefox ESR which will work with Vimperator until May 2018 as the readme states. By then projects will, hopefully, be a bit more stable.

[Vimperator readme]: https://github.com/vimperator/vimperator-labs/blob/master/README.md

**Comment by robehickman** [comment]
@Koushien IMO the 'for security' reasoning is pathetic because there is nothing on that site which is any more/less security sensitive than any other website. It's an interface for installing add ons for pity sake, I'd be far more concerned about Internet banking sites and the like. I know what I'm doing and don't need Mozilla treating me like an idiot and making my browser unuseable in the process.

**Comment by R030t1** [comment]
@robehickman: I agree, and there are a number of misguided changes that Mozilla has undertaken in the past (a particularly bad one was the removal of some extension functionality that a program could get around using the OS). If users here would ask questions on the mailing list about development efforts I think we could affect some positive change.

Something I have found that helps is asking for clear and concise justification for the changes that have been implemented. This should, hopefully, not be burdensome to the developers, as some justification should already exist. If not - then they have admitted to making changes for no reason.

I think it important that the Mozilla developers be held accountable to their users in some way. The changes to extensions are perhaps the least troubling. An important example that I think reflects very poorly on Mozilla is their support for the page visibility API, which YouTube uses to enforce "pause when page not visible." I can't find the exact bug report, but they state that the API shouldn't be disabled because it would subvert user expectations. That's fine, but by that logic Firefox shouldn't block popups by default.

Foundation members seem to be making things up as they go.

**Comment by Hi-Angel** [comment]
> Any other options?

@robehickman please read the scrollback — peoples, me including, have suggested some nice alternatives to Vimperator/Pentadactyl.

**Comment by Hi-Angel** [comment]
@robehickman in fact, you can use [this list on the main page of Qutebrowser](https://github.com/qutebrowser/qutebrowser/#similar-projects), AFAIK the owner tries to keep the list current *(in a way, he writes down any changes he becomes aware of)*.

**Comment by nachoparker** [comment]
pretty happy with vim vixen... they came out of nowhere and it works really well. Also they are pretty active so it seems that they will get the rest working soon.

Some of the things you mention are mozilla's fault, so you won't find a FF addon that supports them.

**Comment by robehickman** [comment]
UI modification can still be done with userChrome.css

/* Get rid of tab bar */
#TabsToolbar {
    visibility: collapse !important;
}

/* Get rid of URL bar */
#nav-bar {
  visibility: collapse !important;
}

**Comment by bovine3dom** [comment]
@robehickman We found that autohiding the UI was a decent compromise - https://github.com/cmcaine/tridactyl/blob/master/src/static/userChrome-minimal.css - as it doesn't break anything.

**Comment by jmercouris** [comment]
I would also like to throw my 2 cents in for those looking for a replacement in the form of a browser, mine is based on GTK and Cocoa (using native OS wherever possible)


a VI compatibility layer is possible to program on top of it, check it out and let me know what you think: https://github.com/nEXT-Browser/nEXT

**Comment by SammysHP** [comment]
I think we can close this issue now as it is clear that Vimperator will never be available as a webextension. Other projects were already mentioned and looking at their GitHub repos is simply amazing. They have more activity in a few weeks than Vimperator during the last years. Thanks for all supporters of Vimperator and please support other projects as well. Eventually there will be an extension as good and rich of features as Vimperator and maybe even better.

**Comment by gkatsev** [comment]
Thanks @SammysHP! Yeah, it's sad to see vimperator go. Maybe someone at some point will resurrect it for one of the firefox forks that are keeping the old addon API. But it's good to see that people are picking up the slack elsewhere!

**Comment by mcarton** [comment]
@gkatsev, @SammysHP and others, in this happy/sad moment, I just wanted to say: thanks for bringing us and maintaining Vimperator all those years! ❤️

**Comment by taocp** [comment]
Thank you all. Thank you Vimperator!

**Comment by maoanran** [comment]
So sad 😭 It is the only reason to use Firefox for many one.

Really thanks for you guys these years. 🥇 🥇 🥇

**Comment by zunsthy** [comment]
Vimperator is No.1.
Thank you all.

**Comment by janzhen** [comment]
It's sad. Vimperator is the only reason I stay with Firefox.
Thank you all.

**Comment by MasahiroSakoda** [comment]
Thank you so much!!
I think Vimperator was one of greatest 
addon.

**Comment by ajwsert** [comment]
I see this has been closed, long discussed and I don't personally use the add-on but I would like to add my two cents here as well.

I see that some has argued against but I would tell to give PaleMoon some more love. I have switched to it so soon I learned about the Mozilla's announcement about dropping XUL and I haven't had any issues with it. Almost all of the add-ons I had been using on Firefox are compatible to some extend and for the most popular ones there are forks to make them fully compatible.

In fact, I went even to the extreme to simply copy my FF's profile files (add-ons files; history, bookmarks, cookies db files etc) to a PM profile and it worked great "out-of-box", I got all my data and add-ons in PM exactly as it was in FF. The only issue was that PM promptly marked some add-ons like incompatible (it doesn't support Jetpack SDK) but around 90% of the add-ons kept working. Even so, some of the incompatible ones could be replaced with PM targeted ones (e.g. Adblock Plus for FF is incompatible, but there is a version from the same dev made for PM [Adblock Latitude]).

Regarding HTML5 or other technologies, I haven't had issues with it. I can't recall any websites not rendering properly in it (maybe 1 or 2 in a thousand) and possibly it was due websites targeting specific browsers through the User Agent string (PM even has a setting to change it to a Gecko or FF compatibility mode). I have used websites with Websockets, HTML5 video etc in it with no issues.

Ok, I have been long already. I just wanted to drop my experience with PM and tell to not give up giving it some attention and who knows this Vimp can live some more in it.

**Comment by 0re5ama** [comment]
Is this project dead?

**Comment by Hi-Angel** [comment]
In some sense. There's no point in further development, given modern Firefox removed all the necessary API. See the discussion for alternatives.

**Comment by 0re5ama** [comment]
Sadly none of the alternatives feels even close to vimperator.

**Comment by dschrempf** [comment]
I can recomment `qutebrowser`, it comes pretty close!

**Comment by nachoparker** [comment]
happy user of Vim Vixen here

**Comment by zouhair** [comment]
Well this is a huge disappointment :(

Thanks for all the good days you gave us, have a good life

**Comment by ZirconCode** [comment]
So long and thanks for all the fish =(

**Comment by tvestelind** [comment]
Vimperator is a great plugin and the only reason I've delayed switching to 57. Thank you for all your work!

**Comment by quba42** [comment]
I have been using an older version of Vimperator (3.13.1) with Pale Moon.
With the last Pale Moon update a couple of extra Vimperator features broke, but for the time being even a partial Vimperator is still preferable to any alternatives I've tried.

If I find the time (a big if), I may look into some more systematic testing of Vimperator features within Pale Moon.
Given the partial "out of the box" compatibility it should be possible to create a Vimperator for Pale Moon fork with reasonable effort...

**Comment by bovine3dom** [comment]
Have you tried Pale Moon's port of Pentadactyl? https://addons.palemoon.org/addon/pentadactyl/

**Comment by dotancohen** [comment]
Vimperator will be missed. It blazed a new trail in UI for a web browser, actually making the web comfortable to use.

Tridactyl walks in Vimperator's footsteps. As of early 2019 I find it feature complete for everyday browsing. Highly recommended to those who mourn the mighty Vimperator.

https://github.com/tridactyl/tridactyl
https://addons.mozilla.org/en-US/firefox/addon/tridactyl-vim/

**Comment by feniksa** [comment]
vimperator is best of the best! Wish to have same functionality in future (currently use tridactyl as alternative).
